Make the `Query` enum a simple struct.

cjgillot opened the pull request:
A lot of code in `rustc_query_system` is generic over it, only to encode an exceptional error case: query cycles.
The delayed computations are now done at cycle detection.

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1678197491) [(plain)](https://github.com/rust-lang/rust/commit/f89c53451df2bcccfdc7aa98e1028749e031ec41/checks/1678197491/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
    Checking rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
    Checking chalk-engine v0.36.0
error[E0423]: expected value, found macro `error`
   --> compiler/rustc_query_system/src/query/plumbing.rs:250:54
    |
250 |                 let error = report_cycle(tcx.sess(), error);
    |                                                      ^^^^^ not a value
    |
help: use `!` to invoke the macro
    |
250 |                 let error = report_cycle(tcx.sess(), error!);

error: aborting due to previous error

For more information about this error, try `rustc --explain E0423`.
For more information about this error, try `rustc --explain E0423`.
error: could not compile `rustc_query_system`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_lint" "-p" "rustc_index" "-p" "rustc_macros" "-p" "rustc_attr" "-p" "rustc_lexer" "-p" "rustc_trait_selection" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_parse_format" "-p" "rustc_errors" "-p" "rustc_lint_defs" "-p" "rustc_save_analysis" "-p" "rustc_feature" "-p" "rustc_parse" "-p" "rustc_span" "-p" "rustc_arena" "-p" "rustc_session" "-p" "rustc_fs_util" "-p" "rustc_middle" "-p" "rustc_query_system" "-p" "rustc_type_ir" "-p" "rustc_apfloat" "-p" "rustc_ast_pretty" "-p" "rustc_serialize" "-p" "rustc_data_structures" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_ast_passes" "-p" "rustc_plugin_impl" "-p" "rustc_ast" "-p" "rustc_interface" "-p" "rustc_incremental" "-p" "rustc_symbol_mangling" "-p" "rustc_typeck" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_ast_lowering" "-p" "rustc_traits" "-p" "rustc_privacy" "-p" "rustc_builtin_macros" "-p" "rustc_resolve" "-p" "rustc_ty_utils" "-p" "rustc_passes" "-p" "rustc_mir_build" "-p" "rustc_target" "-p" "rustc_hir_pretty" "-p" "rustc_error_codes" "-p" "rustc_hir" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:13
```

</details>

cjgillot commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

bors commented:
:hourglass: Trying commit e08e9742ba5989cdb6b27f51226311a2e6372d5a with merge dea0e5624588437b1a358664ff4c776fae7d9625...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e08e9742ba5989cdb6b27f51226311a2e6372d5a","merge_sha":"dea0e5624588437b1a358664ff4c776fae7d9625"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1681925735)
Build commit: dea0e5624588437b1a358664ff4c776fae7d9625 (`dea0e5624588437b1a358664ff4c776fae7d9625`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1681925735"},"merge_sha":"dea0e5624588437b1a358664ff4c776fae7d9625"} -->

rust-timer commented:
Queued dea0e5624588437b1a358664ff4c776fae7d9625 with parent a2cd91ceb0f156cb442d75e12dc77c3d064cdde4, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a2cd91ceb0f156cb442d75e12dc77c3d064cdde4&end=dea0e5624588437b1a358664ff4c776fae7d9625).

@rustbot label: +S-waiting-on-perf

rust-timer commented:
Finished benchmarking try commit (dea0e5624588437b1a358664ff4c776fae7d9625): [comparison url](https://perf.rust-lang.org/compare.html?start=a2cd91ceb0f156cb442d75e12dc77c3d064cdde4&end=dea0e5624588437b1a358664ff4c776fae7d9625).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

cjgillot commented:
@matthewjasper Do you intend to review this or do you want it reassigned?

matthewjasper commented:
This should be reassigned, I'm don't really understand the query system implementation well enough to review this.

Mark-Simulacrum commented:
While I review, let's kick off a more recent perf run - it seems like it's worthwhile to see if some of the mixed results (and minor regressions to bootstrap builds) from the previous one have gone away now. @bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit e08e9742ba5989cdb6b27f51226311a2e6372d5a with merge e332f2214ecfdd09201163667071ec79884b208f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"e08e9742ba5989cdb6b27f51226311a2e6372d5a","merge_sha":"e332f2214ecfdd09201163667071ec79884b208f"} -->

Mark-Simulacrum commented on compiler/rustc_middle/src/ty/query/plumbing.rs:
```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
-                        (queries::$name::describe(tcx, key), stringify!($name))
-                    })*
+        impl Query {
+            $(#[allow(nonstandard_style)] $(#[$attr])*
```
Can we add a comment here that's something like "generate constructors for Query structs for each query"? It took me a bit to understand what was going on here.

cjgillot replied:
Done.

Mark-Simulacrum commented on compiler/rustc_middle/src/ty/query/plumbing.rs:
```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
         #[derive(Clone, Debug)]
-        pub enum Query<$tcx> {
-            $($(#[$attr])* $name($($K)*)),*
+        pub struct Query {
```
I think a comment explaining what this struct is would be helpful. As far as I can tell, it's not really the query - that would imply a single instance for each (statically declared) query. Rather, this represents a particular invocation of the query (with some key), right? It's also an interesting mix of static data (like the name) and the dynamic data (hash, description, span) which we get at runtime from the key(s).

cjgillot replied:
Renamed the struct to `QueryStackFrame` and added a comment.

Mark-Simulacrum commented on compiler/rustc_middle/src/ty/query/plumbing.rs:
```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
-            $($(#[$attr])* $name($($K)*)),*
+        pub struct Query {
+            pub name: &'static str,
+            hash: Fingerprint,
```
Can we explain what this hash is of in a comment?

cjgillot replied:
Done.

Mark-Simulacrum commented on compiler/rustc_middle/src/ty/query/plumbing.rs:
```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
+            pub fn $name<$tcx>(tcx: TyCtxt<$tcx>, key: $($K)*) -> Self {
+                let kind = dep_graph::DepKind::$name;
+                let name = stringify!($name);
+                let description = queries::$name::describe(tcx, key);
```
It looks like calculating the description is actually relatively expensive, and was previously done on demand but is now done for (each? many?) query invocation - is it worth trying to avoid it if we're not going to need it, perhaps by using an Option here and not computing it in the common case?

cc https://github.com/rust-lang/rust/pull/81901 which switches to using String rather than Cow, which was never borrowed anyway.

cjgillot replied:
This function is only called to construct the query backtrace, on exceptional cases: panics and cycles. The number of invocations has not changed, since the number of calls to `describe` used to be exactly the number of instances of the former `Query` enum.

Mark-Simulacrum commented on compiler/rustc_middle/src/ty/query/plumbing.rs:
```diff
@@ -259,52 +256,54 @@ macro_rules! define_queries {
-                    r
-                }
-            }
+                    Some(key.default_span(tcx))
```
This also seems plausibly expensive, FWIW, and presumably not necessary unless we hit an error.

Mark-Simulacrum reviewed (commented):
This seems interesting but at least with the previous performance results not really worth it.

I left some feedback, but it's all mostly minor things and asking for documentation.

I do think it might be worth exploring either reusing or defining a Key trait and storing the Query struct as something roughly like this which might help mitigate some of the performance problems while still giving us a more readable/maintainable implementation.

```
struct Query<'a> {
    name: &'static str,
    key: &'a dyn Key,
}
```

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1858930361)
Build commit: e332f2214ecfdd09201163667071ec79884b208f (`e332f2214ecfdd09201163667071ec79884b208f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1858930361"},"merge_sha":"e332f2214ecfdd09201163667071ec79884b208f"} -->

rust-timer commented:
Queued e332f2214ecfdd09201163667071ec79884b208f with parent 921ec4b3fca17cc777766c240038d7d50ba98e0d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=921ec4b3fca17cc777766c240038d7d50ba98e0d&end=e332f2214ecfdd09201163667071ec79884b208f).

rust-timer commented:
Finished benchmarking try commit (e332f2214ecfdd09201163667071ec79884b208f): [comparison url](https://perf.rust-lang.org/compare.html?start=921ec4b3fca17cc777766c240038d7d50ba98e0d&end=e332f2214ecfdd09201163667071ec79884b208f).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

cjgillot commented:
Rebased on top of https://github.com/rust-lang/rust/pull/81901, new commits (after "Rename to QueryStackFrame") addressing review.

Mark-Simulacrum commented on compiler/rustc_query_system/src/query/mod.rs:
```diff
@@ -25,15 +25,20 @@ use rustc_errors::Diagnostic;
     pub description: String,
     span: Option<Span>,
+    /// This hash is used to deterministically pick
+    /// a query to remove cycles in the parallel compiler.
```
Should this be cfg(parallel_compiler) then? Probably doesn't actually matter in practice since presumably it's not hit at all in successful compilations?

Mark-Simulacrum commented:
@bors try @rust-timer queue

Ok, let's check perf again - thanks for rebasing - the new commits look excellent to me, let's see if the performance is still a bit regressed.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 9ee7c536b7bc0c4fa32c8f528986321a8ae761af with merge 43755cd432ec7823ad91f5cedc4f81457c94ce78...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9ee7c536b7bc0c4fa32c8f528986321a8ae761af","merge_sha":"43755cd432ec7823ad91f5cedc4f81457c94ce78"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1866102334)
Build commit: 43755cd432ec7823ad91f5cedc4f81457c94ce78 (`43755cd432ec7823ad91f5cedc4f81457c94ce78`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1866102334"},"merge_sha":"43755cd432ec7823ad91f5cedc4f81457c94ce78"} -->

rust-timer commented:
Queued 43755cd432ec7823ad91f5cedc4f81457c94ce78 with parent 185de5f41a64e1b790524d5873dc1f7e368f64ab, future [comparison URL](https://perf.rust-lang.org/compare.html?start=185de5f41a64e1b790524d5873dc1f7e368f64ab&end=43755cd432ec7823ad91f5cedc4f81457c94ce78).

rust-timer commented:
Finished benchmarking try commit (43755cd432ec7823ad91f5cedc4f81457c94ce78): [comparison url](https://perf.rust-lang.org/compare.html?start=185de5f41a64e1b790524d5873dc1f7e368f64ab&end=43755cd432ec7823ad91f5cedc4f81457c94ce78).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

Mark-Simulacrum commented:
So the performance still looks overall negative to me - do you have some thoughts on why that might be? I would expect this to either be neutral or slightly positive. Maybe we need a few #[inline] annotations to gain back inlining opportunities lost by removing generics on some functions?

I think this is a worthwhile cleanup and likely outweighs the performance hit, so it's likely I will approve regardless, so if you'd like to squash the fixup commits in that'd be great.

cjgillot commented:
Rebased with some squashing.
I added a few inline directives, but probably not enough.
I made the hashing of the key conditioned on `parallel_compiler`.

bors commented:
:umbrella: The latest upstream changes (presumably #81855) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Mark-Simulacrum commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 6917615c1c3b5d80fd373c3017988c9c51c56602 with merge c3a434afa367e876d114accdcfa5bfc70486bc28...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6917615c1c3b5d80fd373c3017988c9c51c56602","merge_sha":"c3a434afa367e876d114accdcfa5bfc70486bc28"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1942944662)
Build commit: c3a434afa367e876d114accdcfa5bfc70486bc28 (`c3a434afa367e876d114accdcfa5bfc70486bc28`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1942944662"},"merge_sha":"c3a434afa367e876d114accdcfa5bfc70486bc28"} -->

rust-timer commented:
Queued c3a434afa367e876d114accdcfa5bfc70486bc28 with parent e7c23ab933ebc1f205c3b59f4ebc85d40f67d404, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e7c23ab933ebc1f205c3b59f4ebc85d40f67d404&end=c3a434afa367e876d114accdcfa5bfc70486bc28).

rust-timer commented:
Finished benchmarking try commit (c3a434afa367e876d114accdcfa5bfc70486bc28): [comparison url](https://perf.rust-lang.org/compare.html?start=e7c23ab933ebc1f205c3b59f4ebc85d40f67d404&end=c3a434afa367e876d114accdcfa5bfc70486bc28).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

Mark-Simulacrum commented:
Great -- performance looks to me pretty much either neutral or an improvement across the board now. This continues to look great, especially with that last set of changes, @bors r+

bors commented:
:pushpin: Commit 6917615c1c3b5d80fd373c3017988c9c51c56602 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 6917615c1c3b5d80fd373c3017988c9c51c56602 -->
<!-- homu: {"type":"Approved","sha":"6917615c1c3b5d80fd373c3017988c9c51c56602","approver":"Mark-Simulacrum"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #70951) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1943552933) [(plain)](https://github.com/rust-lang/rust/commit/80b9d9340f324d9c3aac0357d97afc88c83890d7/checks/1943552933/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_attr v0.0.0 (/checkout/compiler/rustc_attr)
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
    Checking rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
error[E0405]: cannot find trait `DepKind` in this scope
   --> compiler/rustc_query_system/src/query/job.rs:446:20
    |
446 | fn remove_cycle<D: DepKind>(
    |
help: consider importing this trait
    |
    |
1   | use crate::dep_graph::DepKind;

error: aborting due to previous error

For more information about this error, try `rustc --explain E0405`.
For more information about this error, try `rustc --explain E0405`.
error: could not compile `rustc_query_system`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_codegen_ssa" "-p" "rustc_incremental" "-p" "rustc_graphviz" "-p" "rustc_macros" "-p" "rustc_span" "-p" "rustc_arena" "-p" "rustc_serialize" "-p" "rustc_index" "-p" "rustc_data_structures" "-p" "rustc_errors" "-p" "rustc_lint_defs" "-p" "rustc_target" "-p" "rustc_session" "-p" "rustc_feature" "-p" "rustc_attr" "-p" "rustc_lexer" "-p" "rustc_ast_pretty" "-p" "rustc_apfloat" "-p" "rustc_hir" "-p" "rustc_symbol_mangling" "-p" "rustc_middle" "-p" "rustc_query_system" "-p" "rustc_type_ir" "-p" "rustc_ast" "-p" "rustc_fs_util" "-p" "rustc_driver" "-p" "rustc_plugin_impl" "-p" "rustc_hir_pretty" "-p" "rustc_save_analysis" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_ast_passes" "-p" "rustc_error_codes" "-p" "rustc_parse" "-p" "rustc_lint" "-p" "rustc_trait_selection" "-p" "rustc_infer" "-p" "rustc_parse_format" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_interface" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_typeck" "-p" "rustc_query_impl" "-p" "rustc_privacy" "-p" "rustc_builtin_macros" "-p" "rustc_traits" "-p" "rustc_resolve" "-p" "rustc_passes" "-p" "rustc_mir_build" "-p" "rustc_ty_utils" "-p" "rustc_ast_lowering" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:01:56
```

</details>

Mark-Simulacrum commented:
@bors r+ 

(Feel free to do this yourself too, if you're just rebasing)

bors commented:
:pushpin: Commit ea9100f216e2fa276d9fc84dbc1f06c2e81b2bb8 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum ea9100f216e2fa276d9fc84dbc1f06c2e81b2bb8 -->
<!-- homu: {"type":"Approved","sha":"ea9100f216e2fa276d9fc84dbc1f06c2e81b2bb8","approver":"Mark-Simulacrum"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1943694852) [(plain)](https://github.com/rust-lang/rust/commit/ea9100f216e2fa276d9fc84dbc1f06c2e81b2bb8/checks/1943694852/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_attr v0.0.0 (/checkout/compiler/rustc_attr)
    Checking rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
    Checking rustc_parse v0.0.0 (/checkout/compiler/rustc_parse)
    Checking rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
error[E0405]: cannot find trait `DepKind` in this scope
   --> compiler/rustc_query_system/src/query/job.rs:446:20
    |
446 | fn remove_cycle<D: DepKind>(
    |
help: consider importing this trait
    |
    |
1   | use crate::dep_graph::DepKind;

error: aborting due to previous error

For more information about this error, try `rustc --explain E0405`.
For more information about this error, try `rustc --explain E0405`.
error: could not compile `rustc_query_system`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_serialize" "-p" "rustc_macros" "-p" "rustc_interface" "-p" "rustc_attr" "-p" "rustc_lexer" "-p" "rustc_query_impl" "-p" "rustc_query_system" "-p" "rustc_arena" "-p" "rustc_resolve" "-p" "rustc_codegen_llvm" "-p" "rustc_fs_util" "-p" "rustc_llvm" "-p" "rustc_trait_selection" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_parse_format" "-p" "rustc_incremental" "-p" "rustc_expand" "-p" "rustc_lint_defs" "-p" "rustc_ty_utils" "-p" "rustc_privacy" "-p" "rustc_ast_lowering" "-p" "rustc_mir_build" "-p" "rustc_apfloat" "-p" "rustc_traits" "-p" "rustc_index" "-p" "rustc_symbol_mangling" "-p" "rustc_builtin_macros" "-p" "rustc_typeck" "-p" "rustc_passes" "-p" "rustc_ast_passes" "-p" "rustc_error_codes" "-p" "rustc_hir" "-p" "rustc_data_structures" "-p" "rustc_plugin_impl" "-p" "rustc_target" "-p" "rustc_ast_pretty" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_ast" "-p" "rustc_metadata" "-p" "rustc_parse" "-p" "rustc_hir_pretty" "-p" "rustc_save_analysis" "-p" "rustc_errors" "-p" "rustc_middle" "-p" "rustc_type_ir" "-p" "rustc_lint" "-p" "rustc_span" "-p" "rustc_session" "-p" "rustc_feature" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:11
```

</details>

cjgillot commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 903f65f215812ca4d40e1efc6a5f234e0ab5728c has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 903f65f215812ca4d40e1efc6a5f234e0ab5728c -->
<!-- homu: {"type":"Approved","sha":"903f65f215812ca4d40e1efc6a5f234e0ab5728c","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 903f65f215812ca4d40e1efc6a5f234e0ab5728c with merge c32beb4dd2a6d8e1783ba1b66999d2e1757f2ca5...
<!-- homu: {"type":"BuildStarted","head_sha":"903f65f215812ca4d40e1efc6a5f234e0ab5728c","merge_sha":"c32beb4dd2a6d8e1783ba1b66999d2e1757f2ca5"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc-cargo`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1948704648) [(plain)](https://github.com/rust-lang-ci/rust/commit/c32beb4dd2a6d8e1783ba1b66999d2e1757f2ca5/checks/1948704648/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test test_frequency::prop_frequency_indexed ... ok

failures:

---- test_reverse::prop_reverse_no_headers stdout ----
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-779]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-786]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-792]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-798]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-804]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-810]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-817]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-823]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-830]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-836]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-843]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-849]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-855]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-862]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-869]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-874]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-882]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-887]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-893]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-899]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-905]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-911]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-918]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-924]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-931]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["2埫G"], ["䧄穩:9"], ["~"], ["~턨"], ["x&\u{4}|"], ["r\";"], ["i&0Q"], ["D"], ["‖"], ["zᚚ"], ["B"], [" _"], ["=l?"], ["g¯Q"], [""], ["¯>< "], ["\u{8c})N"], ["2_"], ["\u{4a53a}\u{d2315}_"], [""], ["i5"], ["튢<d"], ["\u{206e}\u{fff6}"], ["\u{99}\u{6dade}톒?"], [""], ["„"], ["]L\u{e}"], ["\u{602}t"], [""], ["\u{81}?"], ["\u{603}\u{6}~!"], ["¨"], [""], [""], [""], ["\u{89}ť"], [""], ["\u{0}>\u{13}"], ["\u{5}\u{95}형&"], [""], [""], ["-Ḳ."], ["\n[\u{c}"], [""], ["뗒"], ["\u{7}:"], [""], ["\u{8d}\u{f433}\u{8a}s"], ["K\u{1d173}"], ["\u{10fffe}9mP"], ["\';:§"], [";\u{c}¬"], [" 핋J#"], ["©^"], [""], ["71X"], ["”i\u{18}"], ["¢#\u{fffb}"], ["3V+:"], ["4"], ["!\u{ea24}"], ["E"], ["+%00"], ["("], ["F~L`"], [")\u{8c}5"], ["\'\u{12}9"], ["]f"], ["\n\u{83}"], ["\u{600}\u{8c}\u{f},"], ["\u{6dd}g"], [""], ["\u{0}#"], ["^"], ["\u{f}_"], ["S"], ["\u{c}狋T\u{89}"], ["B\\"], ["\u{fff3}"], [""], ["\u{11}6\u{1d173}"], ["\u{afc69}⁊"], ["y&4"], ["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["5\u{9f}$"]]`,
 right: `[["2埫G"], ["䧄穩:9"], ["~"], ["~턨"], ["x&\u{4}|"], ["r\";"], ["i&0Q"], ["D"], ["‖"], ["zᚚ"], ["B"], [" _"], ["=l?"], ["g¯Q"], [""], ["¯>< "], ["\u{8c})N"], ["2_"], ["\u{4a53a}\u{d2315}_"], [""], ["i5"], ["튢<d"], ["\u{206e}\u{fff6}"], ["\u{99}\u{6dade}톒?"], [""], ["„"], ["]L\u{e}"], ["\u{602}t"], [""], ["\u{81}?"], ["\u{603}\u{6}~!"], ["¨"], [""], [""], [""], ["\u{89}ť"], [""], ["\u{0}>\u{13}"], ["\u{5}\u{95}형&"], [""], [""], ["-Ḳ."], ["\n[\u{c}"], [""], ["뗒"], ["\u{7}:"], [""], ["\u{8d}\u{f433}\u{8a}s"], ["K\u{1d173}"], ["\u{10fffe}9mP"], ["\';:§"], [";\u{c}¬"], [" 핋J#"], ["©^"], [""], ["71X"], ["”i\u{18}"], ["¢#\u{fffb}"], ["3V+:"], ["4"], ["!\u{ea24}"], ["E"], ["+%00"], ["("], ["F~L`"], [")\u{8c}5"], ["\'\u{12}9"], ["]f"], ["\n\u{83}"], ["\u{600}\u{8c}\u{f},"], ["\u{6dd}g"], [""], ["\u{0}#"], ["^"], ["\u{f}_"], ["S"], ["\u{c}狋T\u{89}"], ["B\\"], ["\u{fff3}"], [""], ["\u{11}6\u{1d173}"], ["\u{afc69}⁊"], ["y&4"], ["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-941]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-947]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-952]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["2埫G"], ["K\u{1d173}"], ["\u{10fffe}9mP"], ["\';:§"], [";\u{c}¬"], [" 핋J#"], ["©^"], [""], ["71X"], ["”i\u{18}"], ["¢#\u{fffb}"], ["3V+:"], ["4"], ["!\u{ea24}"], ["E"], ["+%00"], ["("], ["F~L`"], [")\u{8c}5"], ["\'\u{12}9"], ["]f"], ["\n\u{83}"], ["\u{600}\u{8c}\u{f},"], ["\u{6dd}g"], [""], ["\u{0}#"], ["^"], ["\u{f}_"], ["S"], ["\u{c}狋T\u{89}"], ["B\\"], ["\u{fff3}"], [""], ["\u{11}6\u{1d173}"], ["\u{afc69}⁊"], ["y&4"], ["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["5\u{9f}$"]]`,
 right: `[["2埫G"], ["K\u{1d173}"], ["\u{10fffe}9mP"], ["\';:§"], [";\u{c}¬"], [" 핋J#"], ["©^"], [""], ["71X"], ["”i\u{18}"], ["¢#\u{fffb}"], ["3V+:"], ["4"], ["!\u{ea24}"], ["E"], ["+%00"], ["("], ["F~L`"], [")\u{8c}5"], ["\'\u{12}9"], ["]f"], ["\n\u{83}"], ["\u{600}\u{8c}\u{f},"], ["\u{6dd}g"], [""], ["\u{0}#"], ["^"], ["\u{f}_"], ["S"], ["\u{c}狋T\u{89}"], ["B\\"], ["\u{fff3}"], [""], ["\u{11}6\u{1d173}"], ["\u{afc69}⁊"], ["y&4"], ["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-958]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-965]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-970]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[[""], ["\u{0}#"], ["^"], ["\u{f}_"], ["S"], ["\u{c}狋T\u{89}"], ["B\\"], ["\u{fff3}"], [""], ["\u{11}6\u{1d173}"], ["\u{afc69}⁊"], ["y&4"], ["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["5\u{9f}$"]]`,
 right: `[[""], ["\u{0}#"], ["^"], ["\u{f}_"], ["S"], ["\u{c}狋T\u{89}"], ["B\\"], ["\u{fff3}"], [""], ["\u{11}6\u{1d173}"], ["\u{afc69}⁊"], ["y&4"], ["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-978]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-983]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-989]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["5\u{9f}$"]]`,
 right: `[["_”\u{9b}"], ["4\u{9d}>"], ["逨ࣅ$"], ["\t\u{80}"], [""], ["&"], ["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-993]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1000]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1006]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["5\u{9f}$"]]`,
 right: `[["‖"], [""], ["`G¥"], ["~𥻴\u{14}―"], ["⁓⁔"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1012]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1017]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1023]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["~𥻴\u{14}―"], ["⁓⁔"], ["5\u{9f}$"]]`,
 right: `[["~𥻴\u{14}―"], ["⁓⁔"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1027]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1034]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1039]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["~𥻴\u{14}―"], ["5\u{9f}$"]]`,
 right: `[["~𥻴\u{14}―"], ["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1045]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1050]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1057]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["5\u{9f}$"]]`,
 right: `[["\u{feff}5\u{9f}$"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1061]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1068]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1073]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1079]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[["5"]]`,
 right: `[["\u{feff}5"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1084]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1092]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1094]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1102]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at 'assertion failed: `(left == right)`
  left: `[]`,
 right: `[["\u{feff}"]]`', tests\test_reverse.rs:22:5
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1107]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1113]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1118]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1125]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1129]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1136]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1142]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1147]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1153]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1160]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1165]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1171]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1177]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1184]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1190]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1197]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
[D:\a\rust\rust\build\ct\xsv\target\debug\xit\prop_reverse_no_headers\test-1202]: "D:\\a\\rust\\rust\\build\\ct\\xsv\\target\\debug\\xsv" "reverse" "in.csv" "--no-headers"
thread 'test_reverse::prop_reverse_no_headers' panicked at '[quickcheck] TEST FAILED (runtime error). Arguments: (CsvData { data: [[[239, 187, 191]]] })
Error: "assertion failed: `(left == right)`\n  left: `[]`,\n right: `[[\"\\u{feff}\"]]`"', C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\quickcheck-0.7.1\src\tester.rs:176:28

failures:
    test_reverse::prop_reverse_no_headers

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1948878389)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1948878389","builder_name":"checks-actions"} -->

cjgillot commented:
I can't reproduce the failure.
@bors retry

bors commented:
:hourglass: Testing commit 903f65f215812ca4d40e1efc6a5f234e0ab5728c with merge 301ad8a4fa3ea56fb980443b7997c8f9d72dd717...
<!-- homu: {"type":"BuildStarted","head_sha":"903f65f215812ca4d40e1efc6a5f234e0ab5728c","merge_sha":"301ad8a4fa3ea56fb980443b7997c8f9d72dd717"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1968058538)
Approved by: Mark-Simulacrum
Pushing 301ad8a4fa3ea56fb980443b7997c8f9d72dd717 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1968058538"},"merge_sha":"301ad8a4fa3ea56fb980443b7997c8f9d72dd717"} -->
